### PR TITLE
fix(core): publish transport callbacks before layout effects

### LINKED ISSUE #6421: assistant-transport: useLatestRef publishes from a passive effect
`packages/core/src/react/runtimes/assistant-transport/useLatestRef.ts` publishes with a passive effect:

```ts
export function useLatestRef<T>(value: T) {
  const ref = useRef(value);
  useEffect(() => {
    ref.current = value;
  }, [value]);
  return ref as { current: T };
}
```

`runManager.ts:33-36` wraps `onRun`, `onFinish`, `onCancel`, and `onError` with it, and reads them at call time from inside a `queueMicrotask` (`runManager.ts:51,57,60,66`). so the reads are genuinely cross-render, which is what makes the publication timing load bearing.

passive effects flush child-first within a commit, so a descendant that starts a run from its own `useEffect` (or any `useLayoutEffect` in the tree) runs before this one and calls the previous render's callback. that is the same window the review on #6405 flagged, and #6405 landed on `useInsertionEffect` for exactly this reason. the repo already carries the zero-window form twice, `packages/store/src/useAui.ts:355` with a comment spelling out the ordering, and `packages/tap/src/react-shim/useReactEffectEvent.ts:11`.

worth being precise about scope: `useLatestRef` writes from an effect, not during render, so it does not have the abandoned-render hazard that the rest of this sweep is about. only the stale-window half applies.

`useLatestRef` is not exported from `@assistant-ui/core`, so this is an internal change. the options are switching its body to `useInsertionEffect`, or deleting it in favour of `useEffectEvent` where the value is a callback, which all four call sites are.

<a href="https://rupic.dev/s/s2.BrptICHGmsxnvJchSUcDwDQQr71fK2YRhltfjXHOgiSH5gmra7zBkFnQbWg?ref=github" target="_blank" rel="noopener noreferrer"><picture><source media="(prefers-color-scheme: dark)" srcset="https://rupic.dev/badges/track-dark-v1.svg">Track in Rupic</picture></a>

### PR DESCRIPTION
**Pull request by Kinfe123**
Closes #6421

## Summary

- publish assistant-transport callback refs during the insertion phase
- prevent descendant layout effects from observing callbacks from the previous render
- add a focused timing regression and a patch changeset

## Why

`useLatestRef()` updated its ref from a passive effect. After React committed a render with callback B, a descendant layout effect could schedule transport work before that passive effect ran and still invoke callback A. The assistant-transport run manager uses this helper for `onRun`, `onFinish`, `onCancel`, and `onError`, so the stale callback could belong to a previous configuration or scope.

Publishing the ref from `useInsertionEffect` closes that same-commit window without mutating shared state during render.

## Verification

- reproduced on current main: after rerendering from callback A to callback B, a descendant layout effect observed callback A
- `pnpm --filter @assistant-ui/core test`
- `pnpm lint`
- `pnpm turbo build --filter=@assistant-ui/core...`

<a href="https://cubic.dev/pr/assistant-ui/assistant-ui/pull/6450?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

<a href="https://rupic.dev/s/s2.uLKNxFS01t6HALXxl122VB0SC7PPDtH1JS3kxIN7TtGCZI0jMMT1RllZXgU?ref=github" target="_blank" rel="noopener noreferrer"><picture><source media="(prefers-color-scheme: dark)" srcset="https://rupic.dev/badges/track-dark-v1.svg">Track in Rupic</picture></a>

### CODE REVIEW

File: packages/core/src/react/runtimes/assistant-transport/transport-scheduling.test.ts [hunk 1]
```diff
@@ -85,6 +86,38 @@ const useTransportSchedulingHarness = (
...
+        callbackRef: useLatestRef(callback),
+      });
+
+    const view = render(
+      createElement(Probe, { read: false, callback: callbackA }),
+    );
+    view.rerender(createElement(Probe, { read: true, callback: callbackB }));
+
+    expect(seen).toHaveLength(1);
+    expect(seen[0]).toBe(callbackB);
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/react/runtimes/assistant-transport/transport-scheduling.test.ts]: Good point. I rewrote the regression to drive `useRunManager.schedule()` from a descendant layout effect and assert that the current `onRun` callback is invoked. The test deterministically fails with the old passive-effect publication and passes with `useInsertionEffect`. Full core tests and lint are green.

File: packages/core/src/react/runtimes/assistant-transport/transport-scheduling.test.ts [hunk 2]
```diff
@@ -85,6 +85,52 @@ const useTransportSchedulingHarness = (
 };
 
 describe("assistant transport scheduling contracts", () => {
+  it("uses current callbacks when scheduled by a descendant layout effect", async () => {
+    const onRunA = vi.fn(async () => {});
+    const onRunB = vi.fn(async () => {});
+    const queueMicrotaskSpy = vi
+      .spyOn(globalThis, "queueMicrotask")
+      .mockImplementation((callback) => callback());
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/react/runtimes/assistant-transport/transport-scheduling.test.ts:95]: Agreed. I added the ordering rationale beside the mock and replaced the `waitFor` with a direct assertion. The focused regression and file-level lint/format checks pass.

File: packages/core/src/react/runtimes/assistant-transport/transport-scheduling.test.ts [hunk 3]
```diff
@@ -85,6 +85,54 @@ const useTransportSchedulingHarness = (
...
+      });
+    };
+
+    try {
+      const view = render(
+        createElement(Probe, { enabled: false, onRun: onRunA }),
+      );
+      view.rerender(createElement(Probe, { enabled: true, onRun: onRunB }));
+
+      expect(onRunB).toHaveBeenCalledTimes(1);
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/react/runtimes/assistant-transport/transport-scheduling.test.ts:130]: Valid edge case. I now flush the async run settlement with `await act(async () => {})` before asserting, so the trailing `setIsRunning(false)` stays inside React’s test boundary. The focused regression and file-level lint/format checks pass.
